File: delivery_app/lib/data/models/entry_model.dart
import 'package:equatable/equatable.dart';

class EntryModel extends Equatable {
  final int id;
  final int customerId;
  final String? customerName;
  final String? customerPhone;
  final String? customerAddress;
  final int? deliveryBoyId;
  final String? deliveryBoyName;
  final double milkQuantity;
  final double collectedMoney;
  final int pendingBottles;
  final double rate;
  final String paymentMethod;
  final String? transactionPhoto;
  final bool isDelivered;
  final String? notDeliveredReason;
  final String entryDate;
  final String? createdAt;

  const EntryModel({
    required this.id,
    required this.customerId,
    this.customerName,
    this.customerPhone,
    this.customerAddress,
    this.deliveryBoyId,
    this.deliveryBoyName,
    required this.milkQuantity,
    required this.collectedMoney,
    required this.pendingBottles,
    required this.rate,
    required this.paymentMethod,
    this.transactionPhoto,
    required this.isDelivered,
    this.notDeliveredReason,
    required this.entryDate,
    this.createdAt,
  });

  factory EntryModel.fromJson(Map<String, dynamic> json) {
    return EntryModel(
      id: json['id'],
      customerId: json['customer_id'],
      customerName: json['customer_name'],
      customerPhone: json['customer_phone'],
      customerAddress: json['customer_address'],
      deliveryBoyId: json['delivery_boy_id'],
      deliveryBoyName: json['delivery_boy_name'],
      milkQuantity: double.tryParse(json['milk_quantity'].toString()) ?? 0,
      collectedMoney: double.tryParse(json['collected_money'].toString()) ?? 0,
      pendingBottles: json['pending_bottles'] ?? 0,
      rate: double.tryParse(json['rate'].toString()) ?? 0,
      paymentMethod: json['payment_method'] ?? 'cash',
      transactionPhoto: json['transaction_photo'],
      isDelivered: json['is_delivered'] ?? true,
      notDeliveredReason: json['not_delivered_reason'],
      entryDate: json['entry_date'],
      createdAt: json['created_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'customer_id': customerId,
      'delivery_boy_id': deliveryBoyId,
      'milk_quantity': milkQuantity,
      'collected_money': collectedMoney,
      'pending_bottles': pendingBottles,
      'rate': rate,
      'payment_method': paymentMethod,
      'transaction_photo': transactionPhoto,
      'is_delivered': isDelivered,
      'not_delivered_reason': notDeliveredReason,
      'entry_date': entryDate,
    };
  }

  @override
  List<Object?> get props => [
        id,
        customerId,
        customerName,
        customerPhone,
        customerAddress,
        deliveryBoyId,
        deliveryBoyName,
        milkQuantity,
        collectedMoney,
        pendingBottles,
        rate,
        paymentMethod,
        transactionPhoto,
        isDelivered,
        notDeliveredReason,
        entryDate,
        createdAt,
      ];
}

File: backend/src/models/entry.model.js
const { pool } = require('../config/database');

class EntryModel {
  static async create(data) {
    const result = await pool.query(`
      INSERT INTO entries (
        customer_id, delivery_boy_id, milk_quantity, collected_money,
        pending_bottles, rate, payment_method, transaction_photo,
        is_delivered, not_delivered_reason, entry_date
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
      RETURNING *
    `, [
      data.customer_id,
      data.delivery_boy_id,
      data.milk_quantity || 0,
      data.collected_money || 0,
      data.pending_bottles || 0,
      data.rate || 0,
      data.payment_method || 'cash',
      data.transaction_photo || null,
      data.is_delivered !== undefined ? data.is_delivered : true,
      data.not_delivered_reason || null,
      data.entry_date || new Date().toISOString().split('T')[0]
    ]);
    
    return result.rows[0];
  }

  static async findById(id) {
    const result = await pool.query(`
      SELECT 
        e.*,
        c.name as customer_name,
        c.phone_number as customer_phone,
        db.name as delivery_boy_name
      FROM entries e
      LEFT JOIN customers c ON e.customer_id = c.id
      LEFT JOIN delivery_boys db ON e.delivery_boy_id = db.id
      WHERE e.id = $1
    `, [id]);
    return result.rows[0];
  }

  static async getByCustomer(customerId, startDate = null, endDate = null) {
    let query = `
      SELECT * FROM entries 
      WHERE customer_id = $1
    `;
    
    const params = [customerId];
    let paramCount = 2;

    if (startDate) {
      query += ` AND entry_date >= $${paramCount}`;
      params.push(startDate);
      paramCount++;
    }

    if (endDate) {
      query += ` AND entry_date <= $${paramCount}`;
      params.push(endDate);
      paramCount++;
    }

    query += ' ORDER BY entry_date DESC';

    const result = await pool.query(query, params);
    return result.rows;
  }

  static async getByDeliveryBoy(deliveryBoyId, date = null) {
    let query = `
      SELECT 
        e.*,
        c.name as customer_name,
        c.phone_number as customer_phone,
        c.address as customer_address
      FROM entries e
      LEFT JOIN customers c ON e.customer_id = c.id
      WHERE e.delivery_boy_id = $1
    `;
    
    const params = [deliveryBoyId];

    if (date) {
      query += ' AND e.entry_date = $2';
      params.push(date);
    }

    query += ' ORDER BY e.entry_date DESC, e.created_at DESC';

    const result = await pool.query(query, params);
    return result.rows;
  }

  static async getTodayEntry(customerId) {
    const today = new Date().toISOString().split('T')[0];
    const result = await pool.query(
      'SELECT * FROM entries WHERE customer_id = $1 AND entry_date = $2',
      [customerId, today]
    );
    return result.rows[0];
  }

  static async update(id, data) {
    const fields = [];
    const values = [];
    let paramCount = 1;

    Object.keys(data).forEach(key => {
      if (data[key] !== undefined && key !== 'id') {
        fields.push(`${key} = $${paramCount}`);
        values.push(data[key]);
        paramCount++;
      }
    });

    if (fields.length === 0) return null;

    values.push(id);
    const query = `
      UPDATE entries 
      SET ${fields.join(', ')}
      WHERE id = $${paramCount}
      RETURNING *
    `;

    const result = await pool.query(query, values);
    return result.rows[0];
  }

  static async delete(id) {
    const result = await pool.query(
      'DELETE FROM entries WHERE id = $1 RETURNING *',
      [id]
    );
    return result.rows[0];
  }
}

module.exports = EntryModel;